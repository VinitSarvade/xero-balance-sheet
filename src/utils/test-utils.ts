import {
  format as prettyFormat,
  plugins as prettyFormatPlugins,
} from "pretty-format";
import { http, HttpResponse } from "msw";
import { BalanceSheetReport } from "@/modules/xero/xeroReport.dto";

export function format(value: any) {
  return prettyFormat(value, {
    escapeRegex: true,
    indent: 2,
    plugins: [
      prettyFormatPlugins.ReactTestComponent,
      prettyFormatPlugins.ReactElement,
      prettyFormatPlugins.DOMElement,
      prettyFormatPlugins.DOMCollection,
      prettyFormatPlugins.Immutable,
      prettyFormatPlugins.AsymmetricMatcher,
    ],
    printFunctionName: false,
  });
}

export const BalanceSheetMock = {
  ReportID: "report-id",
  ReportName: "Balance Sheet",
  ReportDate: "2022-01-01",
  ReportTitles: ["Balance Sheet", "Test Company"],
  Rows: [
    {
      RowType: "Header",
      Cells: [{ Value: "" }, { Value: "2023" }, { Value: "2024" }],
    },
    {
      RowType: "Section",
      Title: "Assets",
      Rows: [],
    },
    {
      RowType: "Section",
      Title: "Bank",
      Rows: [
        {
          RowType: "Row",
          Cells: [
            { Value: "My Bank Account" },
            { Value: "126.70" },
            { Value: "99.60" },
          ],
        },
        {
          RowType: "Row",
          Cells: [
            { Value: "Sample Business" },
            { Value: "92911.00" },
            { Value: "92911.00" },
          ],
        },
        {
          RowType: "Row",
          Cells: [
            { Value: "Sample Business 1" },
            { Value: "11039.00" },
            { Value: "11039.00" },
          ],
        },
        {
          RowType: "SummaryRow",
          Cells: [
            { Value: "Total Bank" },
            { Value: "104076.70" },
            { Value: "104049.60" },
          ],
        },
      ],
    },
    {
      RowType: "Section",
      Title: "Liabilities",
      Rows: [],
    },
    {
      RowType: "Section",
      Title: "Current Liabilities",
      Rows: [
        {
          RowType: "Row",
          Cells: [
            { Value: "Car Loan" },
            { Value: "8000.00" },
            { Value: "8000.00" },
          ],
        },
      ],
    },
    {
      RowType: "Section",
      Title: "",
      Rows: [
        {
          RowType: "SummaryRow",
          Cells: [
            { Value: "Total Current Liabilities" },
            { Value: "8000.00" },
            { Value: "8000.00" },
          ],
        },
      ],
    },
  ],
} as BalanceSheetReport;

export function getHandlers() {
  return [
    http.get(`${process.env.BASE_URL}/api/xero/balance-sheet`, () => {
      return HttpResponse.json(BalanceSheetMock);
    }),
  ];
}
