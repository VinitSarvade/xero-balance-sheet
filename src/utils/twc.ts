import { createTwc } from "react-twc";
import { cn } from "./cn";

export const twc = createTwc({ compose: cn });
