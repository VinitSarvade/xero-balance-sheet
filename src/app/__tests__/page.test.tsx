import { expect, describe, it } from "bun:test";
import { render, within } from "@testing-library/react";
import { formatMoney } from "@/utils/money";
import { format } from "@/utils/test-utils";
import Home from "../page";

describe("Home", () => {
  it("renders ReportTitles and ReportTable components", async () => {
    const HomeComponent = await Home();
    const { getByText, getAllByRole, debug } = render(HomeComponent);

    const titles = getAllByRole("heading");
    expect(titles).toHaveLength(2);
    expect(titles[0].textContent).toBe("Balance Sheet");
    expect(titles[1].textContent).toBe("Test Company");

    const tableHeaders = getAllByRole("columnheader");
    expect(tableHeaders).toHaveLength(3);
    expect(tableHeaders[0].textContent).toBeEmpty();
    expect(tableHeaders[1].textContent).toBe("2023");
    expect(tableHeaders[2].textContent).toBe("2024");

    const tableRows = getAllByRole("row");
    expect(tableRows).toHaveLength(12); // including header

    const firstRowColumns = within(tableRows[1]).getAllByRole("cell");
    expect(firstRowColumns).toHaveLength(1);
    expect(firstRowColumns[0].textContent).toBe("Assets");
    expect(firstRowColumns[0].getAttribute("colspan")).toEqual("3");

    const secondRowColumns = within(tableRows[2]).getAllByRole("cell");
    expect(secondRowColumns).toHaveLength(1);
    expect(secondRowColumns[0].textContent).toBe("Bank");
    expect(secondRowColumns[0].getAttribute("colspan")).toEqual("3");

    const thirdRowColumns = within(tableRows[3]).getAllByRole("cell");
    expect(thirdRowColumns).toHaveLength(3);
    expect(thirdRowColumns[0].textContent).toBe("My Bank Account");
    expect(thirdRowColumns[1].textContent).toBe(formatMoney(126.7));
    expect(thirdRowColumns[2].textContent).toBe(formatMoney(99.6));

    const fourthRowColumns = within(tableRows[4]).getAllByRole("cell");
    expect(fourthRowColumns).toHaveLength(3);
    expect(fourthRowColumns[0].textContent).toBe("Sample Business");
    expect(fourthRowColumns[1].textContent).toBe(formatMoney(92911));
    expect(fourthRowColumns[2].textContent).toBe(formatMoney(92911));

    const fifthRowColumns = within(tableRows[5]).getAllByRole("cell");
    expect(fifthRowColumns).toHaveLength(3);
    expect(fifthRowColumns[0].textContent).toBe("Sample Business 1");
    expect(fifthRowColumns[1].textContent).toBe(formatMoney(11039));
    expect(fifthRowColumns[2].textContent).toBe(formatMoney(11039));

    const lastRowColumns = within(tableRows[11]).getAllByRole("cell");
    expect(lastRowColumns).toHaveLength(3);
    expect(lastRowColumns[0].textContent).toBe("Total Current Liabilities");
    expect(lastRowColumns[1].textContent).toBe(formatMoney(8000));
    expect(lastRowColumns[2].textContent).toBe(formatMoney(8000));
  });

  it("matches snapshot", async () => {
    const HomeComponent = await Home();
    const { asFragment } = render(HomeComponent);

    expect(format(asFragment())).toMatchSnapshot();
  });
});
