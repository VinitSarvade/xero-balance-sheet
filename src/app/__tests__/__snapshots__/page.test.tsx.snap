// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Home matches snapshot 1`] = `
"<DocumentFragment>
  <main
    class="m-4 mx-auto min-h-[calc(100vh-2rem)] max-w-4xl rounded-lg bg-white p-8 shadow-lg outline outline-1 outline-gray-200"
  >
    <h1
      class="text-4xl text-center font-semibold"
    >
      Balance Sheet
    </h1>
    <h2
      class="mt-2 text-2xl tracking-tight text-zinc-700 text-center font-semibold"
    >
      Test Company
    </h2>
    <div
      class="mt-8"
    >
      <table
        class="w-full border-collapse"
      >
        <thead>
          <tr>
            <th />
            <th>
              2023
            </th>
            <th>
              2024
            </th>
          </tr>
        </thead>
        <tbody
          class="capitalize"
        >
          <tr>
            <td
              class="py-2"
              colspan="3"
            >
              <div
                class="font-semibold text-2xl"
                data-testid="section-title"
              >
                Assets
              </div>
            </td>
          </tr>
          <tr>
            <td
              class="py-2"
              colspan="3"
            >
              <div
                class="text-xl font-semibold pl-4"
                data-testid="section-title"
              >
                Bank
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="px-2 py-1.5 pl-4"
            >
              My Bank Account
            </td>
            <td
              class="px-2 text-right"
            >
              ₹126.70
            </td>
            <td
              class="px-2 text-right"
            >
              ₹99.60
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="px-2 py-1.5 pl-4"
            >
              Sample Business
            </td>
            <td
              class="px-2 text-right"
            >
              ₹92,911.00
            </td>
            <td
              class="px-2 text-right"
            >
              ₹92,911.00
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="px-2 py-1.5 pl-4"
            >
              Sample Business 1
            </td>
            <td
              class="px-2 text-right"
            >
              ₹11,039.00
            </td>
            <td
              class="px-2 text-right"
            >
              ₹11,039.00
            </td>
          </tr>
          <tr
            class="border-solid border-gray-400 border-y bg-sky-100"
          >
            <td
              class="px-2 py-1.5 pl-4"
            >
              Total Bank
            </td>
            <td
              class="px-2 text-right"
            >
              ₹1,04,076.70
            </td>
            <td
              class="px-2 text-right"
            >
              ₹1,04,049.60
            </td>
          </tr>
          <tr>
            <td
              class="py-2"
              colspan="3"
            >
              <div
                class="font-semibold text-2xl"
                data-testid="section-title"
              >
                Liabilities
              </div>
            </td>
          </tr>
          <tr>
            <td
              class="py-2"
              colspan="3"
            >
              <div
                class="text-xl font-semibold pl-4"
                data-testid="section-title"
              >
                Current Liabilities
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="px-2 py-1.5 pl-4"
            >
              Car Loan
            </td>
            <td
              class="px-2 text-right"
            >
              ₹8,000.00
            </td>
            <td
              class="px-2 text-right"
            >
              ₹8,000.00
            </td>
          </tr>
          <tr>
            <td
              class="py-2"
              colspan="3"
            >
              <div
                class="text-xl font-semibold"
                data-testid="section-title"
              />
            </td>
          </tr>
          <tr
            class="bg-sky-100 text-lg font-semibold uppercase tracking-tight"
          >
            <td
              class="px-2 py-4"
            >
              Total Current Liabilities
            </td>
            <td
              class="px-2 text-right py-4"
            >
              ₹8,000.00
            </td>
            <td
              class="px-2 text-right py-4"
            >
              ₹8,000.00
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </main>
</DocumentFragment>"
`;
